chain/chain/src/doomslug.rs

```diff
@@ -0,0 +1,1001 @@
+            signer,
+            threshold_mode,
+        }
+    }
+
+    /// Returns the `(hash, height)` of the current tip. Currently is only used by tests.
+    pub fn get_tip(&self) -> (CryptoHash, BlockHeight) {
+        (self.tip.block_hash, self.tip.height)
+    }
+
+    /// Returns the largest height for which we have enough approvals to be theoretically able to
+    ///     produce a block (in practice a blocks might not be produceable yet if not enough time
+    ///     passed since it accumulated enough approvals)
+    pub fn get_largest_height_crossing_threshold(&self) -> BlockHeight {
+        self.largest_threshold_height
+    }
+
+    pub fn get_largest_height_with_doomslug_finality(&self) -> BlockHeight {
+        self.largest_ds_final_height
+    }
+
+    pub fn get_largest_skipped_height(&self) -> BlockHeight {
+        self.largest_promised_skip_height
+    }
+
+    pub fn get_largest_endorsed_height(&self) -> BlockHeight {
+        self.largest_endorsed_height
+    }
+
+    pub fn get_timer_height(&self) -> BlockHeight {
+        self.timer.height
+    }
+
+    pub fn get_timer_start(&self) -> Instant {
+        self.timer.started
+    }
+
+    /// Is expected to be called periodically and processed the timer (`start_timer` in the paper)
+    /// If the `cur_time` way ahead of last time the `process_timer` was called, will only process
+    /// a bounded number of steps, to avoid an infinite loop in case of some bugs.
+    /// Processes sending delayed approvals or skip messages
+    ///
+    /// # Arguments
+    /// * `cur_time` - is expected to receive `now`. Doesn't directly use `now` to simplify testing
+    ///
+    /// # Returns
+    /// A vector of approvals that need to be sent to other block producers as a result of processing
+    /// the timers
+    #[must_use]
+    pub fn process_timer(&mut self, cur_time: Instant) -> Vec<Approval> {
+        let mut ret = vec![];
+        for _ in 0..MAX_TIMER_ITERS {
+            let delay = self
+                .timer
+                .get_delay(self.timer.height.saturating_sub(self.largest_ds_final_height));
+            if cur_time >= self.timer.started + delay {
+                if self.timer.height > self.largest_endorsed_height
+                    && self.timer.height > self.tip.height
+                {
+                    self.largest_promised_skip_height =
+                        std::cmp::max(self.timer.height, self.largest_promised_skip_height);
+                    if let Some(approval) = self.create_approval(self.timer.height + 1, false) {
+                        ret.push(approval);
+                    }
+                }
+
+                // Restart the timer
+                self.timer.started += delay;
+                self.timer.height += 1;
+            } else {
+                break;
+            }
+        }
+
+        ret
+    }
+
+    pub fn create_approval(
+        &self,
+        target_height: BlockHeight,
+        is_endorsement: bool,
```

Comment:
I just want to bring this to our attention, that `bool`-based APIs are usually hard to follow when you read/maintain the code, e.g. the use looks like `self.create_approval(self.timer.height + 1, false)` (it is hard to tell what that `false` implies): http://blakesmith.me/2019/05/07/rust-patterns-enums-instead-of-booleans.html / https://www.reddit.com/r/programming/comments/ebxzp8/dont_use_booleans/